### PR TITLE
Samples: Bluetooth: Mesh: light nRF52 DFU support

### PR DESCRIPTION
**Pull request by Andrewpini**
Adds point to point DFU support for nRF52 series, over SMP, to the mesh light sample.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

### CODE REVIEW

**Comment by Andrewpini** [comment]
Will start adding docs soon

**Comment by Balaklaka** [comment]
Remember to add something to release notes.

File: samples/bluetooth/mesh/light/src/smp_dfu.c
Comment (PavelVPV) [inline samples/bluetooth/mesh/light/src/smp_dfu.c:39]: A false schedule may happen if disconnect happens for Mesh Provisioning or Mesh Proxy:
```
Bluetooth initialized
Mesh initialized
SMP BT service advertising successfully started
I: Primary Element: 0x0002
W: opcode 0x2039 status 0x09
SMP BT service advertising failed to start (rc -120)
```
The simplest fix I think would be to check error code of `bt_le_adv_start` against `-EALREADY`. But maybe the more correct way would be to use extended advertising API here and handle this somehow using `struct bt_le_ext_adv_cb` callbacks.

File: samples/bluetooth/mesh/light/src/main.c
Comment (PavelVPV) [inline samples/bluetooth/mesh/light/src/main.c:57]: Could `SYS_INIT` be used here like in `thingy53.c`?

File: samples/bluetooth/mesh/light/Kconfig
Comment (PavelVPV) [inline samples/bluetooth/mesh/light/Kconfig:15]: Aren't these options enabled in overlay file?

File: samples/bluetooth/mesh/light/Kconfig
Comment (PavelVPV) [inline samples/bluetooth/mesh/light/Kconfig:16]: I've no strong opinion on this, but what if we don't enforce 52 series here. Then it should be possible to run this code with nRF53DK. I guess `smp_dfu.c` could be conditionally compiled in if, for example, `CONFIG_MCUMGR_SMP_BT` is enabled.

File: samples/bluetooth/mesh/light/src/main.c
Comment (Andrewpini) [reply samples/bluetooth/mesh/light/src/main.c:57]: It could, but does not work out of the box. Trying to run it with `k_work` makes the advertising fail. Using `k_work_delayable` however with a seconds delay seems to work. Not sure if this is a little to ad hoc solution. What do you think @PavelVPV ?

File: samples/bluetooth/mesh/light/Kconfig
Comment (Andrewpini) [reply samples/bluetooth/mesh/light/Kconfig:15]: Yes, they are. But `depends on` does not pick the option as far as I know, it only makes the compiler deselect the dependent option unless all `depends on` requirements are fulfilled. I choose those dependencies since they are required to build `smp_dfu.c`

File: samples/bluetooth/mesh/light/Kconfig
Comment (Andrewpini) [reply samples/bluetooth/mesh/light/Kconfig:16]: I get your point. The issue is Thingy 53. I wanted to create a clear distinction between the dfu for thingy53 and the other platforms. Picking 52 series was the most easy option (only excluding `nrf5340dk_nrf5340`)

File: samples/bluetooth/mesh/light/src/smp_dfu.c
Comment (Andrewpini) [reply samples/bluetooth/mesh/light/src/smp_dfu.c:39]: So I guess that is an issue in our other sample using this as well then?

Comment (PavelVPV) [reply samples/bluetooth/mesh/light/src/smp_dfu.c:39]: Yes

File: samples/bluetooth/mesh/light/Kconfig
Comment (PavelVPV) [reply samples/bluetooth/mesh/light/Kconfig:15]: But aren't they enabled if use overlay file? How do you see a user should enable dfu: through overlay file or through menuconfig, or somehow else?

File: samples/bluetooth/mesh/light/Kconfig
Comment (Balaklaka) [inline samples/bluetooth/mesh/light/Kconfig:11]: Do we really need this config option. Can we solve it some other way by using some already existing configuration option that is in the overlay-dfu.conf file.

File: samples/bluetooth/mesh/light/src/main.c
Comment (Balaklaka) [inline samples/bluetooth/mesh/light/src/main.c:56]: Could this use CONFIG_MCUMGR_SMP_BT instead? Or will that add thingy:53 too. Or could it then && !thingy:53

**Review by Balaklaka — CHANGES_REQUESTED** [review]

File: samples/bluetooth/mesh/light/Kconfig
Comment (Andrewpini) [reply samples/bluetooth/mesh/light/Kconfig:15]: I guess it is just redundancy. I want the user to use the overlay to enable dfu

Comment (Andrewpini) [reply samples/bluetooth/mesh/light/Kconfig:15]: I can remove it

File: samples/bluetooth/mesh/light/Kconfig
Comment (Andrewpini) [reply samples/bluetooth/mesh/light/Kconfig:11]: Well, if we want all platforms to run this, except thingy53, we would need to exclude `CONFIG_BOARD_THINGY53_NRF5340_CPUAPP
` and `CONFIG_BOARD_THINGY53_NRF5340_CPUAPP_NS` when running the init of smp_dfu. That would work, but either way we turn this it will look a bit ugly if we are not altering the board files for thingy53

File: samples/bluetooth/mesh/light/src/main.c
Comment (Andrewpini) [reply samples/bluetooth/mesh/light/src/main.c:56]: Seee above comment ⬆️

File: samples/bluetooth/mesh/light/CMakeLists.txt
Comment (Andrewpini) [reply samples/bluetooth/mesh/light/CMakeLists.txt:22]: It requires additional configuration. Can we skip that platform for now just not to complicate this task any more? I guess the support for 52 series is what the customers mainly are interested in

File: samples/bluetooth/mesh/light/Kconfig
Comment (Andrewpini) [reply samples/bluetooth/mesh/light/Kconfig:16]: I answered Ingar on this bellow ⬇️

File: samples/bluetooth/mesh/light/CMakeLists.txt
Comment (PavelVPV) [reply samples/bluetooth/mesh/light/CMakeLists.txt:22]: Excluding compilation of `smp_dfu.c` by `CONFIG_BOARD_THINGY53_NRF5340_CPUAPP`/`_NS` maybe a good option. I was also thinking about taking another sample, which doesn't support thingy53, but they are not much in demand.

Comment (Andrewpini) [reply samples/bluetooth/mesh/light/CMakeLists.txt:22]: Using another sample would solve all problems, except that it would be hard for users to discover it unless we slap the feature in their face through documentation page. Using one of the light related samples is most likely to make users aware of this feature.

File: samples/bluetooth/mesh/light/src/smp_dfu.c
Comment (Andrewpini) [reply samples/bluetooth/mesh/light/src/smp_dfu.c:39]: > A false schedule may happen if disconnect happens for Mesh Provisioning or Mesh Proxy:
> 
> ```
> Bluetooth initialized
> Mesh initialized
> SMP BT service advertising successfully started
> I: Primary Element: 0x0002
> W: opcode 0x2039 status 0x09
> SMP BT service advertising failed to start (rc -120)
> ```
> 
> The simplest fix I think would be to check error code of `bt_le_adv_start` against `-EALREADY`. But maybe the more correct way would be to use extended advertising API here and handle this somehow using `struct bt_le_ext_adv_cb` callbacks.

I've tried and successfully added the extended advertiser to smp_dfu.c. Not sure how this can be used to avoid redundant advertisment of smp though. As far as I can see the `struct bt_le_ext_adv_cb` offers a `sent`, `connected` and `scanned` cb. At least the `connected` CB triggers for both device manager app and for the prov/proxy connections in the mesh app.

Comment (Andrewpini) [reply samples/bluetooth/mesh/light/src/smp_dfu.c:39]: Maybe the 'sent` callback can be used 🤔

Comment (Andrewpini) [reply samples/bluetooth/mesh/light/src/smp_dfu.c]: I have made an attempt using the `sent` CB. It seems to work Ok in testing. Review it and see what you think about it

**Review by Balaklaka — APPROVED** [review]

**Review by PavelVPV — CHANGES_REQUESTED** [review]

File: samples/bluetooth/mesh/light/CMakeLists.txt
Comment (Andrewpini) [reply samples/bluetooth/mesh/light/CMakeLists.txt:22]: It also requires additional configurtaion as far as I have seen

File: samples/bluetooth/mesh/light/README.rst
Comment (Andrewpini) [reply samples/bluetooth/mesh/light/README.rst:193]: I did not have to to make it work at least, so it is not a hard requirement for the current solution at least. But maybe it should be?

File: samples/bluetooth/mesh/light/README.rst
Comment (Andrewpini) [reply samples/bluetooth/mesh/light/README.rst:187]: Makes sence, but I'll leave all documentation decisions up to @mia-ko.

File: doc/nrf/links.txt
Comment (Andrewpini) [reply doc/nrf/links.txt:852]: See doc comment bellow ⬇️

File: samples/bluetooth/mesh/light/README.rst
Comment (Andrewpini) [reply samples/bluetooth/mesh/light/README.rst:49]: See doc comment bellow ⬇️

File: samples/bluetooth/mesh/light/CMakeLists.txt
Comment (Andrewpini) [reply samples/bluetooth/mesh/light/CMakeLists.txt:22]: If we want support for 53DK it requires a separate configuration for the network core that allows using the extended advertiser. Since legacy adv at the moment is the default, such a network core configuration should be conditionally added when using the overlay-dfu.conf. I talked to @m-alperen-sener about it, and while it is possible im concerned that it will be a bit convoluted if we want to add support for 53DK now. When we start using extended advertiser by default I think that this won't be a problem anymore.

Comment (PavelVPV) [reply samples/bluetooth/mesh/light/CMakeLists.txt:22]: Ok, lets leave it as is now.

File: samples/bluetooth/mesh/light/README.rst
Comment (PavelVPV) [reply samples/bluetooth/mesh/light/README.rst:193]: Hmm, lets leave this as is.

File: samples/bluetooth/mesh/light/README.rst
Comment (PavelVPV) [reply samples/bluetooth/mesh/light/README.rst:49]: Ok, but IMHO it would be user friendly to add the link so that users don't need to search for this in the documentation.

File: samples/bluetooth/mesh/light/README.rst
Comment (PavelVPV) [reply samples/bluetooth/mesh/light/README.rst:187]: Ok, but I'm not a fun of re-creating such instructions in everywhere. Having them in one place makes easier to support in future if the mobile application logic, or update archive file name changes. If we feel luck of steps in the main guide, then it should be updated instead.

Comment (mia-ko) [inline samples/bluetooth/mesh/light/README.rst:187]:
```suggestion
After the sample is built with the :file:`overlay-dfu.conf` and programmed to your development kit, support for FOTA upgrades is enabled.
See FOTA in Bluetooth mesh (link) for instructions on how to perform FOTA upgrade and initiate the DFU process.
```
I guess this would replace everything else you wrote in this section.

I would recommend saving your detailed steps in case we need them later, i.e. if we want to add some details either in the existing FOTA over BLE guide or in our own mesh guide regarding FOTA upgrades.

**Review by mia-ko — CHANGES_REQUESTED** [review]

File: doc/nrf/releases/release-notes-changelog.rst
Comment (mia-ko) [inline doc/nrf/releases/release-notes-changelog.rst:213]:
```suggestion
    * Point-to-point Device Firmware Update (DFU) support over the Simple Management Protocol (SMP) for supported nRF52 Series development kits.
```

File: samples/bluetooth/mesh/light/README.rst
Comment (mia-ko) [inline samples/bluetooth/mesh/light/README.rst:49]:
```suggestion
   Point-to-point DFU for :ref:`zephyr:thingy53_nrf5340` is supported by default.
   See :ref:`thingy53_app_update` for more information about updating firmware image on :ref:`zephyr:thingy53_nrf5340`.
```
I wonder if it is better to link to https://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-9109/nrf/ug_thingy53.html#updating-firmware-image than to https://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-9109/nrf/ug_thingy53_gs.html#getting-started-with-precompiled-firmware-samples. What do you think?

**Review by mia-ko — CHANGES_REQUESTED** [review]

**Comment by Andrewpini** [comment]
I've added the suggestions from @mia-ko. Need to add the link to the detailed FOTA guide when it is ready. I also moved the init of smp_dfu to the `bt_ready` CB. The reason is that I've tested the implementation for nrf53DK just to see how it should be done, and it was required to move to make it initialize properly on that platform. As soon as we have cleaned up the SMP dfu stuff for all platforms it should be fairly easy to make the same dfu init apply to all platforms.

File: samples/bluetooth/mesh/light/src/main.c
Comment (PavelVPV) [inline samples/bluetooth/mesh/light/src/main.c:59]: :)

**Review by PavelVPV — APPROVED** [review]

**Review by mia-ko — APPROVED** [review]
Approving this, and once it is merged I will edit samples/bluetooth/mesh/light/README.rst to add the missing link to the FOTA in Bluetooth mesh in my PR.

File: samples/bluetooth/mesh/light/README.rst
Comment (ludvigsj) [inline samples/bluetooth/mesh/light/README.rst:190]: I assume this `(link)` is gonna be made into an actual link before this is merged?

**Review by ludvigsj — APPROVED** [review]

File: samples/bluetooth/mesh/light/README.rst
Comment (ludvigsj) [reply samples/bluetooth/mesh/light/README.rst:190]: Ah of course Mia explained this while I was reviewing. Then I guess it's fine

Comment (Andrewpini) [reply samples/bluetooth/mesh/light/README.rst:190]: Mia will fix it 👍

**Pull request merged**
